core/domain/feature_flag_services.py

```diff
@@ -0,0 +1,252 @@
+        Dictionary having key as the feature name and value as the feature
+        flag domain model if present in the storage layer otherwise None.
+    """
+    feature_name_to_feature_flag_model_dict: Dict[str, Optional[
+        feature_flag_domain.FeatureFlag]] = {}
+    feature_models = config_models.FeatureFlagModel.get_multi(
+        feature_names_list)
+
+    for feature_model in feature_models:
+        if feature_model:
+            feature_with_init_settings = registry.Registry.feature_registry[
+                feature_model.id]
+            last_updated = utils.convert_naive_datetime_to_string(
+                feature_model.last_updated)
+            feature_name_to_feature_flag_model_dict[feature_model.id] = (
+                feature_flag_domain.FeatureFlag.from_dict({
+                    'name': feature_with_init_settings.name,
+                    'description': feature_with_init_settings.description,
+                    'feature_stage': feature_with_init_settings.feature_stage,
+                    'force_enable_for_all_users': (
+                        feature_model.force_enable_for_all_users),
+                    'rollout_percentage': feature_model.rollout_percentage,
+                    'user_group_ids': feature_model.user_group_ids,
+                    'last_updated': last_updated
+                })
+            )
+
+    for feature_name in feature_names_list:
+        if feature_name not in feature_name_to_feature_flag_model_dict:
+            feature_name_to_feature_flag_model_dict[feature_name] = None
+
+    return feature_name_to_feature_flag_model_dict
+
+
+def is_feature_flag_enabled(
+    user_id: Optional[str],
+    feature_name: str,
+    feature_flag: Optional[feature_flag_domain.FeatureFlag] = None
+) -> bool:
+    """Returns True if feature is enabled for the given user else False.
+
+    Args:
+        user_id: str|None. The id of the user, if logged-out user then None.
+        feature_name: str. The name of the feature flag that needs to
+            be evaluated.
+        feature_flag: FeatureFlag. The feature flag domain model.
+
+    Returns:
+        bool. True if the feature is enabled for the given user else False.
+    """
+    if feature_flag is None:
+        feature_flag = registry.Registry.get_feature_flag(feature_name)
+    current_server = feature_flag_domain.get_server_mode()
+
+    if (
+        current_server == feature_flag_domain.ServerMode.TEST and
+        feature_flag.feature_stage == feature_flag_domain.ServerMode.DEV.value
+    ):
+        return False
+
+    if (
+        current_server == feature_flag_domain.ServerMode.PROD and
+        feature_flag.feature_stage in (
+            feature_flag_domain.ServerMode.DEV.value,
+            feature_flag_domain.ServerMode.TEST.value)
+    ):
+        return False
+
+    if feature_flag.force_enable_for_all_users:
+        return True
+    if user_id:
+        random_bytes = secrets.token_bytes(16)
```

Comment:
Is this deterministic ... if the same user retrieves the feature flag multiple times will the result always be the same? (There should probably be a backend test that checks this by doing 500 retrievals for the same user.)

Reply:
I have added the test for this part. Also, I have tested the part where if we move from 50% to 60% then the previous users are still able to access the feature.

Reply:
Can you point me to the 50%/60% test? I can't see it in the code.

Reply:
Let's do some math. Suppose we generate the feature flag value for 1000 users at 5%, that means we expect approx 50 to get it turned on. Then we generate it again for those same users at 10% and expect approx 100 to get it turned on. If things work correctly then the ~100 users at the end will definitely include the initial ~50.

Suppose the flag did not work as above, and instead was generated uniformly at random among 1000 users each time. Then, on the second generation at 10%, the probability that we'd hit all the initial users by chance is approx 0.1^50. This is small enough that the test will very likely fail if the generation wasn't cumulative (i.e. the flag for the 10% case includes all users at the 5% level). 

So I think a test structured along the lines of the first paragraph should be safe. You can vary the numbers and the above approach shows you how to do the calculation.

Reply:
Thanks @seanlip, I have added the test for this.